### ISSUE TITLE
Mutexes / external resources

### ISSUE THREAD

**Issue by be5invis** [avajs/ava#2048]
Sometimes we need to limit some of the tests to run serial, or having a maximum parallel jobs, or wait some external resource to be prepared. Therefore introducing such concepts would be helpful for network-oriented tests or stress tests.

Purposed API:

```typescript
test('Some stress test', async t => {
    const handle = await t.resources('stress-test-capacity').acquire(1000);
    // The "stress-test-capacity" is a kind of resource, we need 1000 of them.
    //
    // All resources would be automatically released after the test ends.
    // Alternatively, use handle.release() to manually release a resource.
});
```

**Comment by sindresorhus** [comment]
@be5invis It's not clear what `1000` means in your example. And what does `resources('memory')` actually do? Is `memory` the key? Would be nice if you could elaborate a bit more about the proposed API.

**Comment by be5invis** [comment]
@sindresorhus 

`t.resource` would return a resource manager of a given resource kind (i.e., memory of DB connection), which allocates resource with a given amount.

```typescript
interface TestContext {
    resource(id: string): IResourceManager ;
}

interface IResourceManager {
    // Acquire resource with given amount
    acquire(amount: 1000): Promise<IResourceHandle>;
}

interface IResourceHanlde {
    release(): void;
}
```

**Comment by sindresorhus** [comment]
But what is `1000` and who enforces/controls its usage? How would AVA enforce this? AVA could easily control concurrency, but not memory usage and other things.

**Comment by be5invis** [comment]
@sindresorhus 
`1000` is simply a number, a number representing the resource quantity that the test claimed it will use. It has no link to the actual memory usage. Ah I should name it `stress` instead of `memory`.
Resource managers will take this number, and decide whether to resolve the promise: resolve, wait or throw a error (unable to allocate).

**Comment by novemberborn** [comment]
Within a test file there's plenty of Node.js modules to help you achieve this. I assume you want to coordinate resources across test files. Could you elaborate on that use case?

**Comment by be5invis** [comment]
@novemberborn Yes, coordinate resources across tests.
I have a lib with some stress tests, may take a lot of memory or whatever, so I want to limit the parallelism among them -- they are distributes in multiple files, belonging to multiple sub modules.

**Comment by lo1tuma** [comment]
IMHO there is no need for builtin support in ava. We use [`throat`](https://www.npmjs.com/package/throat) to limit the amount of parallel runs of tests using puppeteer.

**Comment by be5invis** [comment]
@lo1tuma
Does Throat work across Ava's worker processes?

**Comment by lo1tuma** [comment]
@be5invis good point, probably it doesn’t. I have all the puppeteer-based tests in the same file, so that wasn’t an issue.

**Comment by novemberborn** [comment]
I think this is a duplicate of https://github.com/avajs/ava/issues/1366. @be5invis what do you feel about moving the discussion there?

In order to make progress on this we need to flesh out use cases further, and make the argument for why this can't be done by other modules building on top of AVA. Our proposal process is lacking at the moment, so for now bullet points in GitHub comments will suffice.

**Comment by be5invis** [comment]
Well, my problem is slightly different: #1366 is mostly about preparing some resource before a test run, and my proposal is mainly about parallelism control.



________________________________
发件人: Mark Wubben <notifications@github.com>
发送时间: 星期二, 二月 26, 2019 01:18
收件人: avajs/ava
抄送: Belleve Invis; Mention
主题: Re: [avajs/ava] Mutexes / external resources (#2048)


I think this is a duplicate of #1366<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Favajs%2Fava%2Fissues%2F1366&data=02%7C01%7C%7C7deaf904623045af571208d69b452fb8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636867118788726192&sdata=Y3iah80CG3v13zv3%2Fk54wLpdqbu6J9%2FMP%2FS%2FaVbKUDY%3D&reserved=0>. @be5invis<https://nam03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fbe5invis&data=02%7C01%7C%7C7deaf904623045af571208d69b452fb8%7C84df9e7fe9f640afb435aaaaaaaaaaaa%7C1%7C0%7C636867118788736196&sdata=KkHwxSVWiglkg7trfGC96L9QhXf3cx4RBTsAEH121mw%3D&reserved=0> what do you feel about moving the discussion there?

In order to make progress on this we need to flesh out use cases further, and make the argument for why this can't be done by other modules building on top of AVA. Our proposal process is lacking at the moment, so for now bullet points in GitHub comments will suffice.

―

**Comment by novemberborn** [comment]
Yes but what if we can find a solution that works for both? Roughly, if you could start a "resource manager" that can communicate with test workers, you could use that to share random URLs or mutex states.

**Comment by be5invis** [comment]
@novemberborn 
A resource manager could do both, but in general this issue and #1366 are two different aspects. I recommend handle them separately.

**Comment by novemberborn** [comment]
I'm interested in solving the commonalities between those issues, so that specific solutions can be built on top of that.

**Comment by schmod** [comment]
Agreed that this is a slightly different issue than #1366.  

My $0.02 is that a simply string-based mutex (powered by something like [live-mutex](https://www.npmjs.com/package/live-mutex)) would allow for a clean API, and wouldn't require any additional configuration/setup.  Resources would simply be assigned an arbitrary (string) name by the test author, and the mutex broker wouldn't need to know anything about the tests/resources.

The only drawback is that tests wouldn't be able to request "how much" of a resource they consume.

I'm not sure how many real-world scenarios would actually benefit from quantified resource-management.   In my experiences, I'm either trying to isolate a handful of tests that are memory-hogs, or to disallow _any_ concurrent access to a particular file, database, or database table.  

I don't really see many test authors micromanaging the specific amount of memory that a single test requires, nor do I envision many use-cases where a testing database can handle 2 concurrent connections, but not 8.

---

I'd propose two APIs -- one that restricts concurrent access to a resource across all test processes, and another that's only scoped to a local process.

The second (local) kind of mutex would make it far easier to use existing spy/stub libraries (ie. Sinon) with Ava (#1825).  

```ts
test('should limit concurrency', async (t) => {
  await t.mutex('db');
  // or this shorthand for taking out multiple mutexes at once
  await t.mutex(['db', 'lotsOfMemory']);
});
```

```ts
test('using stubs', async (t) => {
  await t.localMutex('userStub');
  const userStub = sinon.stub(userService, 'getUser').returns({ name: 'jane' });
});
```

---

We could also provide a call-signature that allows us to define a "release handler" that runs any code necessary to "clean up" a shared-resource when its mutex is released (either automatically at the end of a test, or by manually calling `handle.release()`)

```ts
test('release handler', async(t) => {
  const [handle, userStub] = await t.mutex(
    'userStub', 
    () => sinon.stub(userService, 'getUser'), 
    (userStub) => userStub.restore()
  );
});
```

This also creates a slightly cleaner/nicer and type-safe alternative to many uses of `before` and `beforeEach` hooks, allowing shared resources to be instead instantiated and obtained via a helper function (rather than relying on some combination of `beforeEach`, `context` and/or hoisted variables).  Conceptually, this would be vaguely reminiscent of the underlying patterns behind React Hooks.

**Comment by novemberborn** [comment]
@schmod I like it!

To clarify, I think "local" should mean the "test file". We could name that `fileMutex`. But I wonder if `mutex` should be scoped to the test file, and then we could have a `globalMutex` to manage mutexes across concurrently executing files?

I like both automatically releasing mutexes, and allowing it to be managed more directly. We could support both I reckon?

**Comment by ronen** [comment]
> I'm not sure how many real-world scenarios would actually benefit from quantified resource-management. ... nor do I envision many use-cases where a testing database can handle 2 concurrent connections, but not 8.

I've got a scenario that would qualify, I think.  I'm essentially doing integration testing on my front-end JS code communicating with a back-end server which I run locally for testing.   I'd like to run my tests in parallel, with the server handling an independent connection from each test.  But I may need to limit the number of simultaneous connections to the server so I don't thrash my local machine.   In this case it's not a question of 2 or 8 concurrent connections but potentially many more: I've currently got ~65 such integration tests, and of course that number keeps growing.

[Careful readers will note that in #2268 I said my backend server doesn't support simultaneous requests.  But that will hopefully be changing soon, in which case I'd like to take advantage of that to parallelize my tests.  But I'm concerned about *too much* parallelism!]

Being able to limit the amount of simultaneous access to a resource feels to me to be analogous to the `--concurrency` CLI option & default limit, which I think nobody debates as an unreasonable thing to have :)

One simple way to handle this might be to have a `--concurrent-tests` CLI option to would limit the total number of tests that are run simultaneously, though that would be overly restrictive if only some of the tests connect to the resource.  Even if `--concurrent-tests` only affect to the number of concurrent tests in a single file, I could get some overall control of.  

Thinking about it more, I think that case actually *is* the same as  #1366, but with concurrency limits.  Will go make a comment there...

**Comment by novemberborn** [comment]
AVA now has experimental shared worker support. Code runs in a worker thread, in the main process, and can communicate with test workers.

The proposals made in this issue could be implemented using this infrastructure. Please join us in https://github.com/avajs/ava/issues/2605.
